### PR TITLE
Razer Lancehead Tournament Edition (Wired)

### PR DESCRIPTION
**Pull request by aelkz**
Manufacturer: Razer
Device Class: 00(>ifc )
Vendor Id: 1532
Product Id: 0060
Number of Interfaces: 3

### CODE REVIEW

**Comment by aelkz** [comment]
Let me know if I need to provide more information or help with some integrated tests. Thank you.

File: driver/razermouse_driver.c [hunk 1]
```diff
@@ -2453,6 +2486,7 @@ static const struct hid_device_id razer_devices[] = {
     { HID_USB_DEVICE(USB_VENDOR_ID_RAZER,USB_DEVICE_ID_RAZER_MAMBA_2012_WIRELESS) },
     { HID_USB_DEVICE(USB_VENDOR_ID_RAZER,USB_DEVICE_ID_RAZER_MAMBA_WIRED) },
     { HID_USB_DEVICE(USB_VENDOR_ID_RAZER,USB_DEVICE_ID_RAZER_MAMBA_WIRELESS) },
+    { HID_USB_DEVICE(USB_VENDOR_ID_RAZER,USB_DEVICE_ID_LANCEHEAD_TE_WIRED) },
```
Comment (z3ntu) [inline driver/razermouse_driver.c]: Should be `USB_DEVICE_ID_RAZER_LANCEHEAD_TE_WIRED`

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 2]
```diff
@@ -93,6 +93,28 @@ class RazerMambaTE(__RazerDeviceBrightnessSuspend):
...
+    USB_PID = 0x0060
+    HAS_MATRIX = True
+    MATRIX_DIMS = [1, 16]
+    METHODS = ['get_firmware', 'get_matrix_dims', 'has_matrix', 'get_device_name', 'get_device_type_mouse', 'get_brightness', 'set_brightness', 'set_wave_effect',
+               'set_static_effect', 'set_spectrum_effect', 'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect',
+               'set_breath_single_effect', 'set_breath_dual_effect', 'set_custom_effect', 'set_key_row', 'max_dpi',
+               'get_dpi_xy', 'set_dpi_xy']
+
+    RAZER_URLS = {
+        "top_img": "https://d1urewwzb2qwii.cloudfront.net/sys-master/root/h5e/h0b/8888236245022",
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: Aren't there images available on the Razer site? I'd rather not have externally hosted images.

**Review by z3ntu — CHANGES_REQUESTED** [review]

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 2 above)
Comment (aelkz) [reply daemon/openrazer_daemon/hardware/mouse.py]: yeah.. tried to look the generated html page and didn't see any reference. Also I have tried several different naming patterns and didn't find any match.. =/
The product ID is: 27310,27309

For example:
https://assets.razerzone.com/eeimages/products/27310/lanceheadte-og-1200x630-v1.png
https://assets.razerzone.com/eeimages/products/27309/ogimage-1200x630.jpg

File: driver/razermouse_driver.c (hunk 1 above)
Comment (aelkz) [reply driver/razermouse_driver.c]: fixed. thanks

**Comment by z3ntu** [comment]
And you tested everything? Then please also add an entry to the readme with the new device.

**Comment by terricain** [comment]
Also add entry to udev file

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 2 above)
Comment (z3ntu) [reply daemon/openrazer_daemon/hardware/mouse.py]: side_img: https://assets.razerzone.com/eeimages/products/27310/lanceheadte-galleryimg-1500x1000-4.png
perspective_img: https://assets.razerzone.com/eeimages/products/27310/lanceheadte-galleryimg-1500x1000-5.png
Didn't find a nice top_img yet

**Comment by dustan9546** [comment]
Any updates?

**Comment by z3ntu** [comment]
@aelkz:
* Add entry to README
* Add entry to udev rules file
* Run autopep8 on your new script

**Comment by ghost** [comment]
Are you still working on this?

**Comment by z3ntu** [comment]
@dustan9546 @LightNeZz Can you test this branch on your computers? If everything is working properly, we can merge it.

**Comment by ghost** [comment]
@z3ntu What do i have to run? I’m new to linux.

**Comment by z3ntu** [comment]
@LightNeZz What Linux distro do you run?

**Comment by LightNeZz** [comment]
@z3ntu I’m using Linux Mint.

**Comment by tk463** [comment]
@z3ntu I'm on a fedora 25 with a Lancehead TE. what do you want us to check ?

**Comment by z3ntu** [comment]
@tk463 On Fedora running
```
sudo dnf install rpm-build rpmdevtools
rpmdev-setuptree
cd ~/rpmbuild/SPECS
curl -o openrazer.spec https://pastebin.com/raw/GXxqSDJ2
spectool -g -R openrazer.spec
sudo dnf -y builddep openrazer.spec
rpmbuild -bb openrazer.spec
sudo dnf install ../RPMS/noarch/*.rpm
```
should do the job.

**Comment by z3ntu** [comment]
Btw Fedora 25 is EOL and you should update to 26 or 27.

**Comment by tk463** [comment]
Tried messing around a bit. Dpi, device_mode and serial are OK. Can't control the lights though but that might come from somthing else.

@dustan9546 @LightNeZz can you test this branch too ?

@z3ntu I know about fedora 25 but it's at my workplace and I can't really upgrade. It's a pain but I'll manage for a bit.

**Comment by z3ntu** [comment]
Thanks for trying. Lights have to work before merging.
@aelkz Would be great if you could get the rest of the functionality working :) But good job so far.

**Comment by ghost** [comment]
@z3ntu I'm using Linux Mint 18.3 Cinnamon

**Comment by dustan9546** [comment]
@tk463 on vacation for 2 more days. Will try it when I get back.

**Comment by dustan9546** [comment]
What do i need to do for Ubuntu?

**Comment by z3ntu** [comment]
```
git clone https://github.com/aelkz/openrazer
cd openrazer
./scripts/build_debs.sh
sudo dpkg -i dist/*.deb
sudo apt install -f
```

**Comment by dustan9546** [comment]
Didn't do anything

**Comment by z3ntu** [comment]
You have to give me more details than "Didn't do anything"...

**Comment by dustan9546** [comment]
Was that supposed to make the Lancehead show up in Polychromatic?

**Comment by z3ntu** [comment]
Did you reboot already after installing the new .deb files?
And yes

**Comment by dustan9546** [comment]
Said "Failed to generate deb files"

**Comment by z3ntu** [comment]
Then look at the log linked in the output and fix the problems mentioned in there (missing dependencies I guess).

**Comment by dustan9546** [comment]
How do I get to the log? I'm new to Linux :P

**Comment by z3ntu** [comment]
You can use your file manager or do `cat $filename` (where you replace $filename with the filename)

**Comment by z3ntu** [comment]
There will be a line like 
```
Failed to generate deb files. Check /tmp/tmp.8ojGhCJy6Z_deb_build_tmp/build.log for more details
```
in the output of the build_debs.sh script and you will have to look at the file linked in your output.

**Comment by dustan9546** [comment]
bash: /tmp/tmp.YZgZr9BYkx_deb_build_tmp/build.log: Permission denied

**Comment by z3ntu** [comment]
Just put this into your terminal: `cat /tmp/tmp.YZgZr9BYkx_deb_build_tmp/build.log`

**Comment by dustan9546** [comment]
Heres the output
dpkg-buildpackage: source package openrazer
dpkg-buildpackage: source version 2.0.0-0ubuntu1
dpkg-buildpackage: source distribution xenial
dpkg-buildpackage: source changed by Luke Horwell <luke@ubuntu-mate.org>
 dpkg-source --before-build razer
dpkg-buildpackage: host architecture amd64
dpkg-source: warning: unknown information field 'Version' in input data in package's section of control info file
dpkg-checkbuilddeps: error: Unmet build dependencies: debhelper (>= 8.0.0) python3-setuptools
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)

**Comment by z3ntu** [comment]
Then execute `sudo apt install debhelper python3-setuptools` and run the build_debs script again

**Comment by dustan9546** [comment]
Do I reboot after that?

**Comment by z3ntu** [comment]
After you have installed the .deb files, yes.

**Comment by dustan9546** [comment]
The Lancehead is showing up, but I can only change DPI, and if I try to change color, it just changes it on my Ornata.

**Comment by dustan9546** [comment]
Any update?

**Comment by aelkz** [comment]
sorry guys, I was away from home last month.
So, I've tried to execute the tests into my Fedora 25 without success.
Tried to bootup a new laptop with Ubuntu Desktop x64 16.04 and stopped working there..

Since I didn't tried all the test procedures above, I'll try to update the information missing and fix everything that was reported here. I'll be pushing the source until this Friday 19.

Cheers.

**Comment by aelkz** [comment]
So, just gathering all info for starting a new test sprint.

**1) Tried to run it into my laptop without success:**
Linux local.server 4.13.16-100.fc25.x86_64 #1 SMP Mon Nov 27 19:52:46 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
   
```
[root@local SPECS]# spectool -g -R openrazer.spec 
[root@local SPECS]# rpmbuild -bb openrazer.spec 
error: File /root/rpmbuild/SPECS/openrazer.spec does not appear to be a specfile.
```
Maybe because, https://pastebin.com/raw/GXxqSDJ2 is dead?

**2) I'll check into ubuntu 16x64 tonight. Need to be sure that all steps are clear to perform the tests:**

Need to check:
- Lights/Colors control
- Anything else?

```
sudo apt install debhelper python3-setuptools

git clone https://github.com/aelkz/openrazer
cd openrazer
./scripts/build_debs.sh
sudo dpkg -i dist/*.deb
sudo apt install -f

# reboot 
# start polychromatic
```

**Comment by z3ntu** [comment]
I might have set the spec to expire after 30 days or so
But from @tk463 we already know
> Tried messing around a bit. Dpi, device_mode and serial are OK. Can't control the lights though but that might come from somthing else.

**Comment by dustan9546** [comment]
@aelkz Still couldn't change the colors.

**Comment by aelkz** [comment]
@z3ntu I have installed everything from scratch using source https://github.com/openrazer/openrazer/archive/v2.1.1.tar.gz and starting from there to track all changes.

Installed all the rpm packages from ~/rpmbuild/RPMS/noarch
and restarted the computer. Everything fine with the installation.

```
================================================================================================================================================
 Package                                       Arch                   Version                                Repository                    Size
================================================================================================================================================
Installing:
 libxdo                                        x86_64                 1:3.20150503.1-2.fc24                  fedora                        40 k
 openrazer-daemon                              noarch                 2.1.1-1.fc25                           @commandline                 117 k
 openrazer-kernel-modules-dkms                 noarch                 2.1.1-1.fc25                           @commandline                  45 k
 openrazer-meta                                noarch                 2.1.1-1.fc25                           @commandline                 5.7 k
 python3-daemonize                             noarch                 2.4.7-1.fc25                           updates                       15 k
 python3-numpy                                 x86_64                 1:1.11.2-1.fc25                        fedora                       3.0 M
 python3-openrazer                             noarch                 2.1.1-1.fc25                           @commandline                  43 k
 python3-setproctitle                          x86_64                 1.1.10-1.fc25                          updates                       21 k
 xautomation                                   x86_64                 1.07-8.fc24                            fedora                        62 k
 xdotool                                       x86_64                 1:3.20150503.1-2.fc24                  fedora                        53 k

Transaction Summary
================================================================================================================================================
Install  10 Packages

```

What I've got so far:

1)
```
[root@local ci]# lsusb | grep '1532:'
Bus 001 Device 002: ID 1532:0060 Razer USA, Ltd 
```

2)  
```
[root@local ci]# lsmod | grep razer
razerkbd               57344  0
```

3) 
```
[root@local ci]# PYTHONPATH="pylib:daemon" python3 -c "from openrazer.client import DeviceManager; a = DeviceManager(); print(a.devices)"
[]
```

4) 
```
[root@local ci]# openrazer-daemon -r -v --log-dir=/tmp/razer_logs
Traceback (most recent call last):
  File "/usr/bin/openrazer-daemon", line 64, in stop_daemon
    import psutil
ImportError: No module named 'psutil'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/openrazer-daemon", line 168, in <module>
    run()
  File "/usr/bin/openrazer-daemon", line 143, in run
    stop_daemon(args)
  File "/usr/bin/openrazer-daemon", line 70, in stop_daemon
    except ModuleNotFoundError:
NameError: name 'ModuleNotFoundError' is not defined
```

**My specs:**
```
[root@local ci]# /usr/bin/python --version
Python 2.7.13

[root@local ci]# uname -a
Linux local.server 4.13.16-100.fc25.x86_64 #1 SMP Mon Nov 27 19:52:46 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```
Is there something that I'm missing here?

**Comment by z3ntu** [comment]
ImportError: No module named 'psutil'

That's missing. Probably in a package called python3-psutil

**Comment by aelkz** [comment]
getting the following error:

```
[root@local openrazer]# openrazer-daemon -r -v --log-dir=/tmp/razer_logs
Process 12246 stopped
[root@local openrazer]# sudo journalctl -b | grep -i razer_mount
Jan 24 09:14:18 local.server razer_mount[1044]: Driver razermouse
Jan 24 09:14:18 local.server razer_mount[1043]: Driver razermouse
Jan 24 09:14:18 local.server razer_mount[1046]: Device_ID 0003:1532:0060.0001
Jan 24 09:14:18 local.server razer_mount[1042]: Driver razermouse
Jan 24 09:14:18 local.server razer_mount[1048]: Device_ID 0003:1532:0060.0002
Jan 24 09:14:18 local.server razer_mount[1049]: Modprobing razermouse
Jan 24 09:14:18 local.server razer_mount[1050]: Modprobing razermouse
Jan 24 09:14:18 local.server systemd-udevd[937]: Process 'razer_mount razermouse 0003:1532:0060.0001' failed with exit code 1.
Jan 24 09:14:18 local.server systemd-udevd[950]: Process 'razer_mount razermouse 0003:1532:0060.0002' failed with exit code 1.
Jan 24 09:14:18 local.server razer_mount[1045]: Device_ID 0003:1532:0060.0003
Jan 24 09:14:18 local.server razer_mount[1059]: Modprobing razermouse
Jan 24 09:14:18 local.server systemd-udevd[941]: Process 'razer_mount razermouse 0003:1532:0060.0003' failed with exit code 1.
Jan 24 09:19:20 local.server razer_mount[6279]: Driver razerkraken
Jan 24 09:19:20 local.server razer_mount[6280]: Device_ID 0003:1532:0504.0004
Jan 24 09:19:20 local.server razer_mount[6282]: Modprobing razerkraken
Jan 24 09:19:20 local.server systemd-udevd[6274]: Process 'razer_mount razerkraken 0003:1532:0504.0004' failed with exit code 1.
Jan 24 10:14:06 local.server razer_mount[12117]: Driver razerkraken
Jan 24 10:14:06 local.server razer_mount[12119]: Device_ID 0003:1532:0504.0005
Jan 24 10:14:06 local.server razer_mount[12121]: Modprobing razerkraken
Jan 24 10:14:06 local.server systemd-udevd[12110]: Process 'razer_mount razerkraken 0003:1532:0504.0005' failed with exit code 1.
[root@local openrazer]# 
```

**Comment by z3ntu** [comment]
What's the output of `sudo dkms install openrazer-driver/1.0.0` ?

**Comment by aelkz** [comment]
```
[root@local rpmbuild]# polychromatic-tray-applet 
ERROR: Daemon isn't running or crashed.
Exception: Could not connect to daemon


 [root@local rpmbuild]# sudo dkms install openrazer-driver/1.0.0
Module openrazer-driver/1.0.0 already installed on kernel 4.14.14-200.fc26.x86_64/x86_64
```

**Comment by aelkz** [comment]
got it working right now. Updated to Fedora26 also.
The test is ongoing.

**Comment by fiercebrake** [comment]
I would be happy to test, if needed. On Fedora 27.

**Comment by dustan9546** [comment]
Is there anything I can do to help?

**Comment by aelkz** [comment]
colors still not working for some reason.
I had to mount a windows host to start gathering pcap's with razer official app and start debugging.
Don't have much time available because of my current schedule, but I'm working on

**Comment by fiercebrake** [comment]
Thanks for your time and effort, Alex.

Let us know if we can be of any assistance.

**Comment by fiercebrake** [comment]
Whatever we can do to help, just let us know

**Comment by zakanybalazs** [comment]
Hi, just tested it with lancehead. It gets recognised in polychromatic, but lighting still can't be changed. tried to install with the v2.1.1.tar.gz but I coulnd't install it. Couldn't access the log tho, so I can't provide anything useful. [running on ubuntu 16.04]
Would like to help testing :)

**Comment by z3ntu** [comment]
I've rebased this branch on master and pushed it to the branch [`feature_lancehead_te_0x0060`](https://github.com/openrazer/openrazer/tree/feature_lancehead_te_0x0060). No functional changes but removed unnecessary stuff.

~~This device probably only has support for the new kind of software based lighting effects. The daemon doesn't support this yet though.~~ Looked through the caps that are hopefully from this device. Looks like a normal packets without the software stuff. Will try getting it to work later.

**Comment by fiercebrake** [comment]
Much appreciated!

**Comment by fiercebrake** [comment]
The device is recognized by the daemon/polychromatic. It seems that's it for now, effects will not work.

It's nice to finally see it recognized and listed though :-D

**Comment by z3ntu** [comment]
I'm gonna close this now as the content is in the branch I have mentioned before. I'll take a look at it in a couple of weeks.

**Pull request closed without merging**